### PR TITLE
add custom switch widget

### LINKED ISSUE #478: custom material switch widget
we're currently using the default material design from gio material for our switch/toggle widgets. This isn't in conformity with our mockup designs and so a custom switch widget in decred material should be created. 

#### Current Implementation
Screenshot 2021-06-18 at 15 14 19

#### Mockup Designs
Screenshot 2021-06-18 at 15 15 12

Screenshot 2021-06-18 at 15 15 21

It should be noted that the colours used should be the same as specified in the mockups.

### PR DESCRIPTION
**Pull request by Sirmorrison**
Fix #478 
This PR implements custom switch widget.

### CODE REVIEW

File: ui/decredmaterial/switch.go [hunk 1]
```diff
@@ -3,16 +3,26 @@
 package decredmaterial
 
 import (
+	"image"
 	"image/color"
 
+	"gioui.org/f32"
+	// "gioui.org/internal/f32color"
```
Comment (beansgum) [inline ui/decredmaterial/switch.go]: Cleanup imports

File: ui/decredmaterial/switch.go [hunk 2]
```diff
@@ -3,16 +3,26 @@
...
-	"gioui.org/widget/material"
+	// "gioui.org/widget/material"
 )
 
 type Switch struct {
-	material.SwitchStyle
+	Enabled    color.NRGBA
+	Disabled   color.NRGBA
+	thumbColor color.NRGBA
+	Switch     *widget.Bool
```
Comment (beansgum) [inline ui/decredmaterial/switch.go]: I guess this holds value for the active state of the switch, a better variable name would be `isChecked.

File: ui/decredmaterial/switch.go [hunk 3]
```diff
@@ -3,16 +3,26 @@
...
+	"gioui.org/op"
+	"gioui.org/op/clip"
+	"gioui.org/op/paint"
 	"gioui.org/unit"
 	"gioui.org/widget"
-	"gioui.org/widget/material"
+	// "gioui.org/widget/material"
 )
 
 type Switch struct {
```
Comment (beansgum) [inline ui/decredmaterial/switch.go:19]: Make the variables in this struct private and add functions to set them.

Comment (beansgum) [inline ui/decredmaterial/switch.go:19]: This widget needs a `Changed()` event for when the enabled state of the switch is changed.

File: ui/decredmaterial/switch.go [hunk 4]
```diff
@@ -28,8 +38,12 @@ type SwitchButtonText struct {
 	selected                           int
 }
 
-func (t *Theme) Switch(swtch *widget.Bool) Switch {
-	return Switch{material.Switch(t.Base, swtch)}
+func (t *Theme) Switch(swtch *widget.Bool) *Switch {
```
Comment (beansgum) [inline ui/decredmaterial/switch.go]: Make this a primitive bool variable, handle the `*widget.Bool` here and provide a `Changed()` function.
```suggestion
func (t *Theme) Switch(isChecked bool) *Switch {
```

File: ui/decredmaterial/switch.go [hunk 5]
```diff
@@ -28,8 +38,12 @@ type SwitchButtonText struct {
...
+	sw := &Switch{
+		Switch: swtch,
+	}
+	sw.Enabled, sw.Disabled, sw.thumbColor = t.Color.Primary, t.Color.InactiveGray, t.Color.Surface
```
Comment (beansgum) [inline ui/decredmaterial/switch.go]: Disabled and inactive states are mixed up here. Add an inactive state for when the switch is inactive and a disabled state for when user interaction is disabled on the switch.

**Review by beansgum — CHANGES_REQUESTED** [review]
Can a disabled state be added to this widget where it'll ignore all user-triggered events?

File: ui/decredmaterial/switch.go (hunk 4 above)
Comment (oshorefueled) [reply ui/decredmaterial/switch.go]: The Switch widget should really just be a skin that accepts an instance of the bool type. This means it shouldn't return an object. Similar to the Button widget that accepts an instance of the Clickable type, the Switch widget should only need to accept an instance of the Bool type.  

To use the switch widget in this current state, we'll need to keep an instance of both the Switch type and the bool type which is quite redundant.

**Review by oshorefueled — CHANGES_REQUESTED** [review]

File: ui/decredmaterial/switch.go [hunk 6]
```diff
@@ -54,6 +67,106 @@ func (t *Theme) SwitchButtonText(i []SwitchItem) *SwitchButtonText {
...
+	return layout.Dimensions{Size: dims}
+}
+
+func (s *Switch) Changed() bool {
+	return s.switchWidget.Changed()
+}
+
+func (s *Switch) IsChecked() bool {
+	if s.disabled {
+		s.SetChecked(false)
```
Comment (beansgum) [inline ui/decredmaterial/switch.go]: Why is this value set in a read-only function? Also it's possible for a switch to be checked while in a disabled state. The disabled state simply means no user interaction is allowed.

**Review by beansgum — CHANGES_REQUESTED** [review]

**Review by beansgum — APPROVED** [review]

**Review by oshorefueled — APPROVED** [review]

**Pull request merged**